File: LeetCode/binaryTree/104.maximum-depth-of-binary-tree.py
#
# @lc app=leetcode id=104 lang=python3
#
# [104] Maximum Depth of Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root is None: return 0
        if ((root.left is None) and (root.right is None)): return 1
        childDepth = max(self.maxDepth(root.left), self.maxDepth(root.right))
        return childDepth+1


# alt: 自顶向下
# // 记录最大深度
# int res = 0;
# # // 记录遍历到的节点的深度
# int depth = 0;

# # // 主函数
# int maxDepth(TreeNode root) {
# 	traverse(root);
# 	return res;
# # }

# # // 二叉树遍历框架
# void traverse(TreeNode root) {
# 	if (root == null) {
# 		return;
# 	}
# 	# // 前序位置
# 	depth++;
#     if (root.left == null && root.right == null) {
#         // 到达叶子节点，更新最大深度
# 		res = Math.max(res, depth);
#     }
# 	traverse(root.left);
# 	traverse(root.right);
# 	# // 后序位置
# 	depth--;
# }
        
# @lc code=end

File: LeetCode_Go/BinaryTree/104.maximum-depth-of-binary-tree.go
/*
 * @lc app=leetcode id=104 lang=golang
 *
 * [104] Maximum Depth of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// Given a binary tree, return the maximum depth of the tree (length of longest path from the root node to deepest leaf node).

// Travserse the Binary Tree. pre-order operation: add 1 to the depth; post-order operation: subtract 1 from the depth; States: `curDepth` dynamically change, `depth` maintains the maximum depth.

// O(n) time | O(n) space

func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	maxDepth, curDepth := 0, 0
	traverse(root, &maxDepth, &curDepth)
	return maxDepth
}

func traverse(root *TreeNode, maxDepth, curDepth *int) {
	if root == nil {
		return
	}
	*curDepth++
	if *curDepth > *maxDepth {
		*maxDepth = *curDepth
	}
	traverse(root.Left, maxDepth, curDepth)
	traverse(root.Right, maxDepth, curDepth)
	*curDepth--
}

// @lc code=end

